client/gc_test.go

```diff
@@ -343,136 +341,53 @@ func TestAllocGarbageCollector_MakeRoomForAllocations_GC_Fallback(t *testing.T)
 	}
 }
 
-// TestAllocGarbageCollector_MaxAllocs asserts that when making room for new
+// TestAllocGarbageCollector_MakeRoomFor_MaxAllocs asserts that when making room for new
 // allocs, terminal allocs are GC'd until old_allocs + new_allocs <= limit
-func TestAllocGarbageCollector_MaxAllocs(t *testing.T) {
-	t.Skip("Test fails, unsure if behaviour change or broken test")
-
+func TestAllocGarbageCollector_MakeRoomFor_MaxAllocs(t *testing.T) {
 	t.Parallel()
-	server, serverAddr := testServer(t, nil)
-	defer server.Shutdown()
-	testutil.WaitForLeader(t, server.RPC)
-
-	const maxAllocs = 6
-	client, cleanup := TestClient(t, func(c *config.Config) {
-		c.GCMaxAllocs = maxAllocs
-		c.GCDiskUsageThreshold = 100
-		c.GCInodeUsageThreshold = 100
-		c.GCParallelDestroys = 1
-		c.GCInterval = time.Hour
-
-		c.RPCHandler = server
-		c.Servers = []string{serverAddr}
-		c.ConsulConfig.ClientAutoJoin = new(bool) // squelch logs
-	})
-	defer cleanup()
-	waitTilNodeReady(client, t)
-
-	callN := 0
-	assertAllocs := func(expectedAll, expectedDestroyed int) {
-		// Wait for allocs to be started
-		callN++
-		client.logger.Info(fmt.Sprintf("[TEST] %d -- Waiting for %d total allocs, %d GC'd", callN, expectedAll, expectedDestroyed))
-		testutil.WaitForResult(func() (bool, error) {
-			all, destroyed := 0, 0
-			for _, ar := range client.getAllocRunners() {
-				all++
-				if ar.IsDestroyed() {
-					destroyed++
-				}
-			}
-			return all == expectedAll && destroyed == expectedDestroyed, fmt.Errorf(
-				"expected %d allocs (found %d); expected %d destroy (found %d)",
-				expectedAll, all, expectedDestroyed, destroyed,
-			)
-		}, func(err error) {
-			client.logger.Info(fmt.Sprintf("[TEST] %d -- FAILED to find %d total allocs, %d GC'd!", callN, expectedAll, expectedDestroyed))
-			t.Fatalf("%d alloc state: %v", callN, err)
-		})
-		client.logger.Info(fmt.Sprintf("[TEST] %d -- Found %d total allocs, %d GC'd!", callN, expectedAll, expectedDestroyed))
-	}
-
-	// Create a job
-	state := server.State()
-	job := mock.Job()
-	job.TaskGroups[0].Tasks[0].Driver = "mock_driver"
-	job.TaskGroups[0].Tasks[0].Config = map[string]interface{}{
-		"run_for": "30s",
-	}
-	nodeID := client.Node().ID
-	if err := state.UpsertJob(98, job); err != nil {
-		t.Fatalf("error upserting job: %v", err)
-	}
-	if err := state.UpsertJobSummary(99, mock.JobSummary(job.ID)); err != nil {
-		t.Fatalf("error upserting job summary: %v", err)
-	}
-
-	newAlloc := func() *structs.Allocation {
-		alloc := mock.Alloc()
-		alloc.JobID = job.ID
-		alloc.Job = job
-		alloc.NodeID = nodeID
-		return alloc
-	}
-
-	// Create the allocations
-	allocs := make([]*structs.Allocation, 7)
-	for i := 0; i < len(allocs); i++ {
-		allocs[i] = newAlloc()
-	}
+	logger := testlog.HCLogger(t)
+	statsCollector := &MockStatsCollector{}
+	conf := gcConfig()
+	conf.MaxAllocs = 3
+	gc := NewAllocGarbageCollector(logger, statsCollector, &MockAllocCounter{}, conf)
```

Comment:
I am not convinced this test is actually testing what the old test was. Using the MockAllocCounter is essentially skipping the part of the code trying to be tested: https://github.com/hashicorp/nomad/blob/d7dbf14c463d3e68022ad28b52b92425bed5d760/client/gc.go#L241

NumAllocs will always be 0